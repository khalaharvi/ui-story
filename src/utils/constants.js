export const PRIMARY_CATEGORY_KEY = 'categories.lvl0';
export const SECONDARY_CATEGORY_KEY = 'categories.lvl1';
export const TERTIARY_CATEGORY_KEY = 'categories.lvl2';
