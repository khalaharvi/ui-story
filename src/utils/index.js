export { default as createClassNames } from './createClassNames';
export { default as getDisplayName } from './getDisplayName';
export { default as capitalizeFirstLetter } from './capitalizeFirstLetter';
export { default as slugify } from './slugify';
