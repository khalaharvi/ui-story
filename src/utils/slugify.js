export default (text) => {
  // Use hash map for special characters
  const specialChars = {"à":'a',"ä":'a',"á":'a',"â":'a',"æ":'a',"å":'a',"ë":'e',"è":'e',"é":'e', "ê":'e',"î":'i',"ï":'i',"ì":'i',"í":'i',"ò":'o',"ó":'o',"ö":'o',"ô":'o',"ø":'o',"ù":'o',"ú":'u',"ü":'u',"û":'u',"ñ":'n',"ç":'c',"ß":'s',"ÿ":'y',"œ":'o',"ŕ":'r',"ś":'s',"ń":'n',"ṕ":'p',"ẃ":'w',"ǵ":'g',"ǹ":'n',"ḿ":'m',"ǘ":'u',"ẍ":'x',"ź":'z',"ḧ":'h',"·":'-',"/":'-',"_":'-',",":'-',":":'-',";":'-'}; // eslint-disable-line

  return text.toString().toLowerCase()
    .replace(/\s+/g, '-') // Replace spaces with -
    .replace(/./g, target => specialChars[target] || target) // Replace special characters using the hash map
    .replace(/&/g, '-and-') // Replace & with 'and'
    .replace(/[^\w-]+/g, '') // Remove all non-word chars
    .replace(/--+/g, '-') // Replace multiple - with single -
    .replace(/^-+/, '') // Trim - from start of text
    .replace(/-+$/, ''); // Trim - from end of text
};
