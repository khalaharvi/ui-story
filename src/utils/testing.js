import { object } from 'prop-types';
import { channel, createBroadcast } from 'emotion-theming';
import { shallow, mount } from 'enzyme';
import theme from '@dering-hall/dh-theme';

const getOptions = (options = {}) => {
  const broadcast = createBroadcast(theme);

  return {
    context: {
      [channel]: broadcast
    },
    childContextTypes: {
      [channel]: object
    },
    ...options
  };
};

export const mountWithTheme = (
  Component,
  options = {}
) => mount(Component, { ...getOptions(options) });

export const shallowWithTheme = (
  Component,
  options = {}
) => shallow(Component, { ...getOptions(options) });

export default {
  mountWithTheme,
  shallowWithTheme
};
