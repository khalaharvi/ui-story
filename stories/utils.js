/* eslint-disable import/prefer-default-export */
import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { InstantSearch } from 'react-instantsearch/dom';
import { connectStateResults } from 'react-instantsearch/connectors';
import { withMedia, MediaQueryProvider } from 'react-media-query-hoc';
import { withTests as withTestsAddon } from '@storybook/addon-jest';
import { ThemeProvider } from 'emotion-theming';
import dhTheme from '@dering-hall/dh-theme';
import testsResults from '../.jest-test-results.json';

const Wrap = ({ index, children }) => (
  <InstantSearch
    appId={process.env.ALGOLIA_ID}
    apiKey={process.env.ALGOLIA_API_KEY}
    indexName={index}
  >
    {children}
  </InstantSearch>
);

Wrap.propTypes = {
  index: PropTypes.string,
  children: PropTypes.node
};

Wrap.defaultProps = {
  index: process.env.ALGOLIA_PRODUCT_INDEX
};

const WrapThemeProvider = ({ children, theme }) => (
  <ThemeProvider theme={theme}>
    <Fragment>
      {children}
    </Fragment>
  </ThemeProvider>
);

WrapThemeProvider.propTypes = {
  theme: PropTypes.object // eslint-disable-line
};

WrapThemeProvider.defaultProps = {
  theme: dhTheme
};

const SearchStateDebug = connectStateResults(({ searchState }) => (
  <div style={{ margin: '1rem 0', fontFamily: 'monospace' }}>
    <pre
      style={{
        background: '#f6f8fa',
        fontSize: '.65rem',
        padding: '.5rem'
      }}
    >
      <strong>searchState</strong> = {JSON.stringify(searchState, null, '\t')}
    </pre>
  </div>
));

const customQueries = {
  small: 'screen and (max-width: 730px)',
  medium: 'screen and (max-width: 970px)',
  large: 'screen and (max-width: 1240px)',
  xlarge: 'screen and (min-width: 1240px)'
};

const MediaQueriesWrapper = ({ children }) => (
  <MediaQueryProvider queries={customQueries}>
    {children}
  </MediaQueryProvider>
);

const OnSmallOnly = withMedia(({ media, children }) => {
  if (!media.small) {
    return null;
  }

  return (
    <Fragment>
      {children}
    </Fragment>
  );
});

const withTests = withTestsAddon({
  results: testsResults
});

export {
  SearchStateDebug,
  MediaQueriesWrapper,
  OnSmallOnly,
  Wrap,
  WrapThemeProvider,
  withTests
};
/* eslint-enable */
