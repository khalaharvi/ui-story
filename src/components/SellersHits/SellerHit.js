import React from 'react';
import PropTypes from 'prop-types';

import { capitalizeFirstLetter } from 'utils';
import Image from 'components/Image';
import { cx } from './SellersHits';

const getSellerTypeDisplayName = types =>
  types
    .map(type =>
      type.split('-').map(typeChunk =>
        capitalizeFirstLetter(typeChunk)).join(' '))
    .join(', ');

// TO DO: add fallback images
const SellerHit = ({ hit, transformHitImageUrl }) => (
  <div to={hit.product_path} className={cx('list-item-link')}>
    <div className={cx('list-item-img')}>
      <Image
        src={
          transformHitImageUrl ?
          transformHitImageUrl(hit.primary_image_url) :
          hit.primary_image_url
        }
        alt={hit.name}
      />
    </div>
    <b className={cx('list-item-name')}>{hit.name}</b>
    <i className={cx('list-item-types')}>{getSellerTypeDisplayName(hit.seller_types)}</i>
  </div>
);

SellerHit.propTypes = {
  transformHitImageUrl: PropTypes.func,
  hit: PropTypes.shape({
    slug: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    seller_types: PropTypes.arrayOf(PropTypes.string).isRequired,
    primary_image_url: PropTypes.string.isRequired,
    primary_type: PropTypes.string.isRequired
  })
};

export default SellerHit;
