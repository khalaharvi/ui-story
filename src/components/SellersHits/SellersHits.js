import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { connectInfiniteHits, connectStateResults } from 'react-instantsearch/connectors';

import { createClassNames } from 'utils';
import Button from 'components/Button';
import SellerHit from './SellerHit';

export const cx = createClassNames('sellers-hits');

// TO DO: add fallback images
export const SellersHits = ({
  className,
  hits,
  hasMore,
  refine,
  searching,
  sellerHitComponent: SellerHitComponent
}) => (
  <div className={classnames(cx(''), className)}>
    <ul className={cx('list')}>
      {hits.map(hit => (
        <li key={hit.objectID} className={cx('list-item')}>
          <SellerHitComponent hit={hit} />
        </li>
      ))}
    </ul>
    <Button
      className={cx('load-more')}
      disabled={searching}
      hidden={!hasMore}
      loading={searching}
      onClick={() => refine()}
      secondary
    >
      {searching ? 'Loading...' : 'Load More'}
    </Button>
  </div>
);

SellersHits.propTypes = {
  sellerHitComponent: PropTypes.func,
  className: PropTypes.string,
  searching: PropTypes.bool.isRequired,
  hasMore: PropTypes.bool.isRequired,
  refine: PropTypes.func.isRequired,
  hits: PropTypes.arrayOf(PropTypes.shape({
    objectID: PropTypes.string.isRequired,
    slug: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    seller_types: PropTypes.arrayOf(PropTypes.string).isRequired,
    primary_image_url: PropTypes.string.isRequired
  }))
};


SellersHits.defaultProps = {
  sellerHitComponent: props => <SellerHit {...props} />
};

export default connectInfiniteHits(connectStateResults(SellersHits));
