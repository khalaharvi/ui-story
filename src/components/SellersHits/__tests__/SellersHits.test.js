import React from 'react';
import { shallow } from 'enzyme';
import { mountWithTheme } from 'utils/testing';

import { SellersHits } from '../SellersHits';

describe('<SellersHits />', () => {
  it('renders correctly', () => {
    const hits = [
      {
        objectID: '123',
        name: 'Designer Name',
        primary_image_url: 'https://placehold.it/370x220',
        slug: '/designer-name',
        seller_types: ['architect'],
        primary_type: 'architect'
      }
    ];

    const wrapper = shallow(
      <SellersHits
        refine={() => null}
        hits={hits}
        searching={false}
        hasMore
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('renders correctly with className', () => {
    const hits = [
      {
        objectID: '123',
        name: 'Designer Name',
        primary_image_url: 'https://placehold.it/370x220',
        slug: '/designer-name',
        seller_types: ['architect'],
        primary_type: 'architect'
      }
    ];

    const wrapper = shallow(
      <SellersHits
        refine={() => null}
        hits={hits}
        searching={false}
        hasMore
        className="custom-products-hits"
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should call refine when the load more button is clicked', () => {
    const refine = jest.fn();
    const hits = [
      {
        objectID: '123',
        name: 'Designer Name',
        primary_image_url: 'https://placehold.it/370x220',
        slug: '/designer-name',
        seller_types: ['interior-designer'],
        primary_type: 'interior-designer'
      },
      {
        objectID: '456',
        name: 'Designer Name 2',
        primary_image_url: 'https://placehold.it/370x220',
        slug: '/designer-name-2',
        seller_types: ['architect'],
        primary_type: 'architect'
      },
      {
        objectID: '789',
        name: 'Designer Name 3',
        primary_image_url: 'https://placehold.it/370x220',
        slug: '/designer-name-3',
        seller_types: ['brand'],
        primary_type: 'brand'
      }
    ];

    const wrapped = mountWithTheme(
      <SellersHits
        hits={hits}
        refine={refine}
        hasMore
        searching={false}
      />
    );

    expect(refine.mock.calls).toHaveLength(0);
    wrapped.find('button').simulate('click');
    expect(refine.mock.calls).toHaveLength(1);
  });

  it('should disable load more button when searching', () => {
    const hits = [
      {
        objectID: '123',
        name: 'Designer Name',
        primary_image_url: 'https://placehold.it/370x220',
        slug: '/designer-name',
        seller_types: ['architect'],
        primary_type: 'architect'
      }
    ];

    const wrapped = mountWithTheme(
      <SellersHits
        refine={() => null}
        hits={hits}
        searching
        hasMore
      />
    );

    expect(wrapped.find('button').props().disabled).toBe(true);
  });

  it('should hide load more button when it is the last page', () => {
    const hits = [
      {
        objectID: '123',
        name: 'Designer Name',
        primary_image_url: 'https://placehold.it/370x220',
        slug: '/designer-name',
        seller_types: ['architect'],
        primary_type: 'architect'
      }
    ];

    const wrapped = mountWithTheme(
      <SellersHits
        refine={() => null}
        hits={hits}
        hasMore={false}
        searching={false}
      />
    );
    expect(wrapped.find('button').props().hidden).toBe(true);
  });
});
