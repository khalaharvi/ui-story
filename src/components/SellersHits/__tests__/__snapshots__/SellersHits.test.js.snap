// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SellersHits /> renders correctly 1`] = `
<div
  className="dh-sellers-hits"
>
  <ul
    className="dh-sellers-hits-list"
  >
    <li
      className="dh-sellers-hits-list-item"
      key="123"
    >
      <sellerHitComponent
        hit={
          Object {
            "name": "Designer Name",
            "objectID": "123",
            "primary_image_url": "https://placehold.it/370x220",
            "primary_type": "architect",
            "seller_types": Array [
              "architect",
            ],
            "slug": "/designer-name",
          }
        }
      />
    </li>
  </ul>
  <ButtonComponent
    className="dh-sellers-hits-load-more"
    disabled={false}
    hidden={false}
    iconAfter={false}
    inverted={false}
    loading={false}
    onClick={[Function]}
    primary={false}
    secondary={true}
    white={false}
  >
    Load More
  </ButtonComponent>
</div>
`;

exports[`<SellersHits /> renders correctly with className 1`] = `
<div
  className="dh-sellers-hits custom-products-hits"
>
  <ul
    className="dh-sellers-hits-list"
  >
    <li
      className="dh-sellers-hits-list-item"
      key="123"
    >
      <sellerHitComponent
        hit={
          Object {
            "name": "Designer Name",
            "objectID": "123",
            "primary_image_url": "https://placehold.it/370x220",
            "primary_type": "architect",
            "seller_types": Array [
              "architect",
            ],
            "slug": "/designer-name",
          }
        }
      />
    </li>
  </ul>
  <ButtonComponent
    className="dh-sellers-hits-load-more"
    disabled={false}
    hidden={false}
    iconAfter={false}
    inverted={false}
    loading={false}
    onClick={[Function]}
    primary={false}
    secondary={true}
    white={false}
  >
    Load More
  </ButtonComponent>
</div>
`;
